staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go

```diff
@@ -196,41 +198,58 @@ func (o *APIResourceOptions) RunAPIResources(cmd *cobra.Command, f cmdutil.Facto
 		}
 	}
 
-	if o.NoHeaders == false && o.Output != "name" {
-		if err = printContextHeaders(w, o.Output); err != nil {
-			return err
-		}
-	}
-
 	sort.Stable(sortableResource{resources, o.SortBy})
-	for _, r := range resources {
+	if o.Output == "yaml" || o.Output == "json" {
+		var data []byte
+		var err error
 		switch o.Output {
-		case "name":
-			name := r.APIResource.Name
-			if len(r.APIGroup) > 0 {
-				name += "." + r.APIGroup
-			}
-			if _, err := fmt.Fprintf(w, "%s\n", name); err != nil {
+		case "yaml":
+			data, err = yaml.Marshal(resources)
+		case "json":
+			data, err = json.Marshal(resources)
```

Comment:
Don't, we have built-in printers for both json and yaml, you just need to wire them properly here and they should be capable of printing all k8s API resource properly.

Reply:
@soltysh I might have missed something, but after delving around in the code trying to do this, the only existing machinery that I could find that implements printers for JSON and YAML is the `genericclioptions.JSONYamlPrintFlags` type. The `ToPrinter` method of that type returns a `printers.ResourcePrinter` implementation, which takes a `runtime.Object` as an argument.

AIUI (probably wrongly!) the object that would be printed by the `apiresources` command is not itself a k8s API object, so it won't implement `runtime.Object`, so it can't be printed by the above printing facility.

There's also the (arguably minor) issue that using the `JSONYamlPrintFlags` type also adds the `--show-managed-fields` flag which doesn't seem relevant to the `api-resources` subcommand and could be misleading to appear in its flag list.

Would you prefer this code to implement a custom type that implements `runtime.Object` in order that it can use the flag logic in `JSONYamlPrintFlags`, or is there some other part of the code that would be better to use?